clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1146,6 +1204,53 @@ void KernelObjVisitor::visitNthArrayElement(const CXXRecordDecl *Owner,
             .Handler...);
 }
 
+template <typename ParentTy, typename... HandlerTys>
+void KernelObjVisitor::visitRecord(const CXXRecordDecl *Owner, ParentTy &Parent,
+                                   const CXXRecordDecl *Wrapper,
+                                   QualType RecordTy,
+                                   HandlerTys &... Handlers) {
+  if (RecordTy->getAsRecordDecl()->hasAttr<SYCLRequiresDecompositionAttr>()) {
+    visitComplexRecord(Owner, Parent, Wrapper, RecordTy, Handlers...);
+  } else {
+    if (!AllTrue<HandlerTys::VisitInsideSimpleContainers...>::Value)
+      visitSimpleRecord(
+          Owner, Parent, Wrapper, RecordTy,
+          HandlerFilter<!HandlerTys::VisitInsideSimpleContainers, HandlerTys>(
+              Handlers)
+              .Handler...);
+
+    if (AnyTrue<HandlerTys::VisitInsideSimpleContainers...>::Value)
+      visitComplexRecord(
+          Owner, Parent, Wrapper, RecordTy,
+          HandlerFilter<HandlerTys::VisitInsideSimpleContainers, HandlerTys>(
+              Handlers)
+              .Handler...);
+  }
+}
+
+template <typename... HandlerTys>
+void KernelObjVisitor::visitArray(const CXXRecordDecl *Owner, FieldDecl *Field,
+                                  QualType ArrayTy, HandlerTys &... Handlers) {
+
+  if (Field->hasAttr<SYCLRequiresDecompositionAttr>()) {
+    visitComplexArray(Owner, Field, ArrayTy, Handlers...);
+  } else {
+    if (!AllTrue<HandlerTys::VisitInsideSimpleContainers...>::Value)
```

Comment:
Could you explain this (and the next if) logic, please?

Reply:
So same as in visitRecord, but, this is a case where we don't think the array needs decomposition.  There are two types of 'handlers' here, ones that have Opted-out of VisitInsideSimpleContainers (meaning, it simply needs to call the 'simple' version), and those who still want to VisitInsideSimpleContainers (and thus need to continue recursing, like visitComplexArray/visitComplexStruct).

We could simply omit these two AllTrue/AnyTrue checks, but the idea is that if NONE of the handlers have Opted-out (or all have for the 2nd), we don't bother calling the VisitSimpleArray/VisitComplexArray, since noone cares.

AnyTrue's ::value member is a bool that contains whether ANY of the handlers have VisitInsideSimpleContainers (via pack-expansion).  AllTrue does what its name implies, contains 'true' for its 'bool' if ALL of the values are true.

In this case, we want to call visitSimple IIF at least 1 hasn't opted out (so not all are 'true').

When we get C++17, these can both become `if constexpr`, but until then, we have to branch on a bool (or at least, hope the optimizer does it for us :) ).